File: Solution_3.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        r = len(matrix)
        c = len(matrix[0])
        startR, startC = 0, c - 1
        # print(r,c)
        while startR < r and startC >= 0:
            if target == matrix[startR][startC]:
                return True
            elif target > matrix[startR][startC]:
                startR += 1
            else:
                startC -= 1
        return False

File: solution3.java
// Time Complexity : O(log N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : no
// Any problem you faced while coding this : no


class Solution {
    public int search(ArrayReader reader, int target) {
        int low = 0, high = 1;

        // Increase high exponentially until target is less than reader.get(high)
        while (reader.get(high) < target) {
            low = high;
            high = 2 * high;
        }

        // Standard binary search within the range [low, high]
        while (low <= high) {
            int mid = low + (high - low) / 2;  // Corrected mid calculation
            if (reader.get(mid) == target) {
                return mid;
            } else if (reader.get(mid) > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return -1;  // Return -1 if the target is not found
    }
}